internal/cms/signed.go

```diff
@@ -0,0 +1,324 @@
+		return nil, ErrNotSignedData
+	}
+
+	var signedData SignedData
+	if _, err := asn1.Unmarshal(contentInfo.Content.Bytes, &signedData); err != nil {
+		return nil, SyntaxError{Message: "invalid signed data", Detail: err}
+	}
+
+	if signedData.Version != 3 {
+		return nil, SyntaxError{Message: fmt.Sprintf("unsupported signed data version: got %d, want 3", signedData.Version)}
+	}
+
+	certs, err := x509.ParseCertificates(signedData.Certificates.Bytes)
+	if err != nil {
+		return nil, SyntaxError{Message: "failed to parse X509 certificates from signed data", Detail: err}
+	}
+
+	return &ParsedSignedData{
+		Content:      signedData.EncapsulatedContentInfo.Content,
+		ContentType:  signedData.EncapsulatedContentInfo.ContentType,
+		Certificates: certs,
+		CRLs:         signedData.CRLs,
+		SignerInfos:  signedData.SignerInfos,
+	}, nil
+}
+
+// Verify attempts to verify the content in the parsed signed data against the signer
+// information. The `Intermediates` in the verify options will be ignored and
+// re-contrusted using the certificates in the parsed signed data.
+// If more than one signature is present, the successful validation of any signature
+// implies that the content in the parsed signed data is valid.
+// On successful verification, the list of signing certificates that successfully
+// verify is returned.
+// If all signatures fail to verify, the last error is returned.
+//
+// References:
+//   - RFC 5652 5   Signed-data Content Type
+//   - RFC 5652 5.4 Message Digest Calculation Process
+//   - RFC 5652 5.6 Signature Verification Process
+//
+// WARNING: this function doesn't do any revocation checking.
+func (d *ParsedSignedData) Verify(ctx context.Context, opts x509.VerifyOptions) ([][]*x509.Certificate, error) {
+	if len(d.SignerInfos) == 0 {
+		return nil, ErrSignerInfoNotFound
+	}
+	if len(d.Certificates) == 0 {
+		return nil, ErrCertificateNotFound
+	}
+
+	intermediates := x509.NewCertPool()
+	for _, cert := range d.Certificates {
+		intermediates.AddCert(cert)
+	}
+	opts.Intermediates = intermediates
+	verifiedSignerMap := map[string][]*x509.Certificate{}
+	var lastErr error
+	for _, signerInfo := range d.SignerInfos {
+		signingCertificate := d.GetCertificate(signerInfo.SignerIdentifier)
+		if signingCertificate == nil {
+			lastErr = ErrCertificateNotFound
+			continue
+		}
+
+		certChain, err := d.VerifySigner(ctx, &signerInfo, signingCertificate, opts)
+		if err != nil {
+			lastErr = err
+			continue
+		}
+
+		thumbprint, err := hashutil.ComputeHash(crypto.SHA256, signingCertificate.Raw)
+		if err != nil {
+			lastErr = err
+			continue
+		}
+		verifiedSignerMap[hex.EncodeToString(thumbprint)] = certChain
+	}
+	if len(verifiedSignerMap) == 0 {
+		return nil, lastErr
+	}
+
+	verifiedSigningCertChains := make([][]*x509.Certificate, 0, len(verifiedSignerMap))
+	for _, certChain := range verifiedSignerMap {
+		verifiedSigningCertChains = append(verifiedSigningCertChains, certChain)
+	}
```

Comment:
Should we move this check inside `for _, signerInfo := range d.SignerInfos {` loop?